### DISCUSSION TITLE
Add calories to energy sensor device class

### DISCUSSION THREAD

**Discussion by dudanov** [home-assistant/architecture#1116] [Entity Models] [answered]
Hello.

This discussion is my first in this project and concerns classes of sensor devices. I wrote a custom integration that allowed to connect fitness equipment to Home Assistant, which needs to create new classes of sensor devices.

I was advised to discuss the proposed changes one at a time. Actually, here is my first thought.

Currently, `SensorDeviceClass.ENERGY` only includes `Joules` and `Wh`. Moreover, apparently for historical reasons, in the `EnergyConverter` class the `kWh` is taken as the normalized unit, which is not the base unit of energy in `SI`.

My proposal is the following:

1. Add a range of calorie energy units: `cal`, `kcal`, `Mcal`, `Gcal`.
2. In the `EnergyConverter`, change the normalized unit from `kWh` to the `SI` base unit `Joule`. This will improve code readability and understanding of multiplication factors for unit conversions.

Thanks in advance.

**Comment by frenck** [comment]
Hi there @dudanov 👋 

Thanks for your proposal. We generally only add common things like this, if multiple integrations available in (and shipped with) Core today can benefit from this. Can you provide those?

Thanks! 👍 

../Frenck

**Reply by dudanov** [reply]
Hi @frenck 

Thanks for the answer.

I propose to supplement the energy class with additional energy units (calories). I can’t say what Core integrations they use, but they are applicable for thermal energy meters. This unit of energy is connected in one way or another with heating water and human energy.

**Reply by joostlek** [reply]
Please try to do due diligence on the core code to find out what integrations can use this. If we don't know how many integrations can use this or there are too little, it will most likely be declined

**Reply by dudanov** [reply]
Calories are used in the following Core integrations: `fitbit`, `overkiz`, `tractive`, `withings`.

**Reply by dudanov** [reply]
In general, I cannot understand why adding a **_standard_** unit of measurement to an _**existing**_ class of sensor devices is considered a significant change. In my opinion, this is an expansion of functionality and standardization, when there is a base of basic standard units of measurement for various types. Users will use them without having to find workarounds and implement them themselves.

**Reply by frenck** [reply]
Thanks for adding those references @dudanov; that is helpful 👍 

> I cannot understand why adding a standard unit of measurement to an existing class of sensor devices is considered a significant change

Significant is, for this specific case, debatable, I agree. However, it does have an impact on potentially all core (and thus also custom) integrations, and we do want to follow the process.

I get your response, but it is not as complex or difficult has you might think. This proposal looks good to me, and not a big deal. We'll have a quick talk/vote about it in the next architecture meeting (I've added this proposal to the agenda) and I'll get back to you after that.

../Frenck

**Comment by emontnemery** [comment, marked as answer]
We've discussed this in the core team.

The suggestion to add `cal`, `kcal`, `Mcal`, `Gcal` as energy units is approved, please go ahead and open a PR with this change.
The proposed refactoring of `EnergyConverter` to change the base unit should happen in separate PRs, where an initial PR would remove the unused `NORMALIZED_UNIT` from the `BaseUnitConverter` class

**Reply by dudanov** [reply]
Thanks for the reply.

I created 2 PRs:
1. The [first one](https://github.com/home-assistant/core/pull/122796#issue-2435884869) adds calories. I also included the refactoring of the base unit to Joules. Otherwise, there would be a need for a chain of consecutive PRs.
2. [PR](https://github.com/home-assistant/core/pull/122797#issue-2435932133) to remove the unused `NORMALIZED_UNIT` class variable.

Yes, but what about integrations that use calories? Should I change them myself or will the authors of the integrations be sent a newsletter?

**Reply by gjohansson-ST** [reply]
On the PR you're making to update the information on the developers site also include a blog post which describes the change. This is information to developers that they can/should act on it.

**Reply by dudanov** [reply]
> On the PR you're making to update the information on the developers site also include a blog post which describes the change. This is information to developers that they can/should act on it.

Thanks. I have [added the post](https://github.com/home-assistant/developers.home-assistant/pull/2262) to the blog.

**Reply by dudanov** [reply]
Hi @frenck 👋
I have all the PRs ready. Is there any hope that they will be included in 2024.8.0b1?

**Reply by joostlek** [reply]
No, it's a new feature, not a bugfix

**Reply by dudanov** [reply]
> No, it's a new feature, not a bugfix

Extending an existing class is a new feature that takes a month and a half to release? That's interesting.
In that case, I need to change the commit with the blog post from 2024.8 to 2024.9 ?

**Reply by joostlek** [reply]
Yes, that's how our release schedule works https://www.home-assistant.io/faq/release/

**Reply by dudanov** [reply]
According to the schedule, today there are exactly 2 weeks left until the beta release. Apparently, the changes discussed will not be included in this release either?..
